onnxruntime/core/common/threadpool.cc

```diff
@@ -0,0 +1,180 @@
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/platform/threadpool.h"
+
+#ifdef USE_EIGEN_THREADPOOL
+#if defined(_MSC_VER)
+#pragma warning(disable : 4267)
+#endif
+#include <unsupported/Eigen/CXX11/ThreadPool>
+#else
+#include "task_thread_pool.h"
+#endif
+
+namespace onnxruntime {
+
+namespace concurrency {
+#ifdef USE_EIGEN_THREADPOOL
+
+// TODO: This is temporarily taken from Eigen until we upgrade its version.
+// Barrier is an object that allows one or more threads to wait until
+// Notify has been called a specified number of times.
+class Barrier {
+ public:
+  Barrier(unsigned int count) : state_(count << 1), notified_(false) {
+    eigen_assert(((count << 1) >> 1) == count);
+  }
+  ~Barrier() {
+    eigen_assert((state_ >> 1) == 0);
+  }
+
+  void Notify() {
+    unsigned int v = state_.fetch_sub(2, std::memory_order_acq_rel) - 2;
+    if (v != 1) {
+      eigen_assert(((v + 2) & ~1) != 0);
+      return;  // either count has not dropped to 0, or waiter is not waiting
+    }
+    std::unique_lock<std::mutex> l(mu_);
+    eigen_assert(!notified_);
+    notified_ = true;
+    cv_.notify_all();
+  }
+
+  void Wait() {
+    unsigned int v = state_.fetch_or(1, std::memory_order_acq_rel);
+    if ((v >> 1) == 0) return;
+    std::unique_lock<std::mutex> l(mu_);
+    while (!notified_) {
+      cv_.wait(l);
+    }
+  }
+
+ private:
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::atomic<unsigned int> state_;  // low bit is waiter flag
+  bool notified_;
+};
+
+class ThreadPool::Impl : public Eigen::NonBlockingThreadPool {
+ public:
+  Impl(const std::string& name, int num_threads)
+      : Eigen::NonBlockingThreadPool(num_threads) {}
+
+  void ParallelFor(int32_t total, std::function<void(int32_t)> fn) {
+    // TODO: Eigen supports a more efficient ThreadPoolDevice mechanism
+    // We will simply rely on the work queue and stealing in the short term.
+    Barrier barrier(static_cast<unsigned int>(total));
+    std::function<void(int32_t)> handle_iteration;
+    handle_iteration = [=, &handle_iteration, &barrier, &fn](int iteration) {
+      fn(iteration);
+      barrier.Notify();
+    };
+
+    for (int32_t id = 0; id < total; ++id) {
+      Schedule([=, &handle_iteration]() { handle_iteration(id); });
+    }
+
+    barrier.Wait();
+  }
+
+  void ParallelForRange(int64_t first, int64_t last, std::function<void(int64_t, int64_t)> fn) {
+    // TODO: Eigen supports a more efficient ThreadPoolDevice mechanism
+    // We will simply rely on the work queue and stealing in the short term.
+    Barrier barrier(static_cast<unsigned int>(last-first+1));
+    std::function<void(int64_t, int64_t)> handle_range;
+    handle_range = [=, &handle_range, &barrier, &fn](int64_t first, int64_t last) {
+      fn(first, last);
+      barrier.Notify();
+    };
+
+    for (int64_t id = first; id <= last; ++id) {
+      Schedule([=, &handle_range]() { handle_range(id, id + 1); });
```

Comment:
Given that the workload passed to each thread of the threadpool is identical wont a round robin based schedule function be as good as the current implementation of eigen? I am asking for opinion.

Reply:
Only in the case where there is a single invocation and single operator with parallel work executing concurrently.  Otherwise, the workload will contain chunks of similar work, but asymmetry between chunks.